### ISSUE TITLE
Uncaught Error Bug When Running Models Locally

### ISSUE THREAD

**Issue by lovelysandlonelys** [conrad-blucher-institute/semaphore#391]
![Image](https://github.com/user-attachments/assets/55b7f41e-8d8b-4230-a726-d6024f9647ef)

Not entirely sure what this is notating, model testing still ran all the way through despite this message. 

Joy: I narrowed it down to this line of code in NDFD_EXP "NDFD_Predictions: NDFDPredictions[str, str] = NDFDPredictions(url, response)".

**Comment by lovelysandlonelys** [comment]
![Image](https://github.com/user-attachments/assets/9809e03b-d04b-42b0-8095-73f76d9917c9)

with slightly more observation while we do other things we think this might be an NDFD issue?

**Comment by abeasley1722** [comment]
Add the line the errorr occurs in in logs

**Comment by matdenkas** [comment]
Steps to reproduce bug:

1. Executing a model using the NDFD ingestion class over and over until error appears

OR

1. Use the generated URL `https://digital.weather.gov/xml/sample_products/browser_interface/ndfdXMLclient.php?whichClient=NDFDgen&lat=27.83444444&lon=-97.06777778&product=time-series&XMLformat=DWML&begin=2024-08-26T20%3A00%3A00&end=2024-08-27T11%3A00%3A00&Unit=m&wspd=wspd&Submit=Submit` Spam refresh until bug appears.


This seems either an intentional slow down, or a collision of requests on NDFDs server
![Image](https://github.com/user-attachments/assets/549b208e-cadc-4d65-a05d-9f16f4c8159c)

**Comment by matdenkas** [comment]
Im not sure how to move forwards with this ticket without a refactor. This is NOT a semaphore bug!

**Comment by matdenkas** [comment]
Seems like we aren't the only ones, and it has to do with this SOAP service https://www.unidata.ucar.edu/mailing_lists/archives/ldm-users/2024/msg00067.html

**Comment by matdenkas** [comment]
It seems we are hitting the REST API. I tried switching over to the SOAP API and it seemed to fair a lot better, but then I think the server fully crashed and now any request I send it crashes.

**Comment by matdenkas** [comment]
**Solution ideas:**
1. A retry system, either for Semaphore as a whole or for just NDFD. Retry a request if failed multiple times pausing between requests.
2. Switch over to the SOAP API
3. Add some times pause or request queue-ing.

**Comment by matdenkas** [comment]
![Image](https://github.com/user-attachments/assets/3fd5d780-af3d-4c27-be8f-7b73a0f0d99d)

I throws 200 on an error

**Comment by matdenkas** [comment]
[PR#420](https://github.com/conrad-blucher-institute/semaphore/pull/420) Contains proper logging and detection of this error

**Comment by ccftissot** [comment]
> ![Image](https://github.com/user-attachments/assets/3fd5d780-af3d-4c27-be8f-7b73a0f0d99d)
> 
> It throws 200 when we get this error

yeah, I am not surprised.  Something we should report to them to fix.  At the very least they should send back a property error code maybe a 503 or at least a 500.

**Comment by ccftissot** [comment]
> **Solution ideas:**
> 
>     1. A retry system, either for Semaphore as a whole or for just NDFD. Retry a request if failed multiple times pausing between requests.
> 
>     2. Switch over to the SOAP API
> 
>     3. Add some times pause or request queue-ing.

I want to make sure I understand the problem correctly. We (Semaphore) are **not** rapidly firing off request to the NDFD Rest API server.  But because the server is popular, it may be receiving a lot of requests and when that happens, our request fails.  Is this what we are talking about?  @matdenkas can you confirm?

My preferred course of action would be to do a little additional research to find out a bit more about the SOAP vs Rest API options.  The post we found on the unidata board was about the soap service it seems.  So maybe it's not the most stable either (especially since @matdenkas  was able to completely crash it! ;-).

I am going to assign this story to me to do the research but in the meantime can someone answer whether we are following the guidelines that Beto found when calling the service?  If we are not, we should discuss. 

My preference would be to limit the fix for now to the NDFD ingestion class, but we need a bit more research to figure out what is a good workaround.

**Comment by matdenkas** [comment]
@ccftissot 

So best case scenario, were hitting them too fast. For things like wind composition (when this bug is most observed) we hit it back to back for wind speed and wind direction. Semaphore sees those two series as completely different things with no knowledge of one another. 

At worst case scenario, it could be a lot worse. Semaphore is executed over multiple processes in parallel, meaning multiple models using NDFD can be executed at the same time. If I remember right we have taken steps to make sure models with longer lead times are executed first, as they contain the data of models with shorter lead times, but then it becomes a race condition. 

 As for doing a retry for NDFD, I do think that's a good step forwards, though I would prefer ofc to do it for everything all at once, but im happy to keep the scope small if you feel thats better. I do think the retry timer is going to need to be >=30 seconds, as it can take some time for NDFD to succeed in getting requests again.

Finally, would it just be better to reach out to them and give them our use case and see what they advise?

**Comment by ccftissot** [comment]
@matdenkas, @lovelysandlonelys , @bestrada33 , @abeasley1722 

After thinking about this some more, here's my take on it.

For the immediate need, I think we should fall back on using the grib2 files generated by the NWS.  The good news is that we already have code in production that download the files and extracts wind speed and direction (among other things) from these grib2 files.  Niall and I wrote that code (in Python) when we revived the TPW cold stunning website a few years back.  I can repurpose and repackage some of that code for Semaphore, it shouldn't take long.  The current code grabs the data from the grib2 files and generates text files using a specific format that we needed, but I can make it generate files with any format we want. Then we can adjust the NDFD ingestion class to read from these files.  I can work with Son and Niall to make sure that where these files are being downloaded is read-accessible by semaphore.  

The reason that I am not a fan of retries and delays is that it is really just a band-aid.  We can put 2 retires at 30 sec interval but (1) if the problem gets worse on their end, then that will just break us again, and (2) that potentially will make it worse on their end.  The GRIB2 files are on an FTP server and we have never had any issues with them, so I think that approach is our best choice right now.

For (potentially way) down the road with Semaphore, I think that at some point, we will need to refactor Semaphore to be a little less unconcerned with resources.  I think we did it the correct way for version 1.0 but in a future version, we should consider refactoring the input gatherer and possibly the series provider in such a way that they can "bunch up" some of these requests.  We may even consider whether more of an event-based/subscriber-publisher model might be appropriate here.  

We can discuss as a team on Friday, but if you all are OK with using the GRIB2 files, I can take this story moving forward and get the GRIB2 file solution in place.  Let me know.

**Comment by matdenkas** [comment]
@ccftissot 

I agree that retries and waits will likely exacerbate the problem, but I also remember from my last job that that was considered fairly slandered for hitting API's, is that not correct? I understand we shouldn't in this situation but would it be better for other ingestion classes?

Pulling from GRIB2 sounds fine enough to me, I am happy to move forwards when you are ready. Are you wanting to add that or send it to me to add?

**Comment by ccftissot** [comment]
I will get working on it.  I will go ahead and close this ticket and create new tickets for moving forward with the GRIB2 files.  I also started documenting NDFD and how to access that from it on the wiki (https://github.com/conrad-blucher-institute/semaphore/wiki/NDFD-Overview).  I'll document how to get data out of GRIB2 file on the wiki as well as I get the code in shape for Semaphore.

**Comment by matdenkas** [comment]
Sorry for getting back to you late. Thank you for the explanation. My last job did lots of things the wrong way, so thanks for taking the time to explain 😭

**Comment by ccftissot** [comment]
Closing this issue as won't fix.  Will create new tickets for implementing code to get data from NDFD grib files.
